test/cpp/jit/test_backend_compiler_lib.cpp

```diff
@@ -110,6 +110,15 @@ namespace {
 // Put here for demonstration of backend
 // as a whole piece. It's used when compilation is required. A dummy function
 // can be passed when there's no usage of compilation in runtime backend lib.
+#ifdef DELEGATE_RUNTIME
+// Runtime does not use preprocess(). Implement a dummy function to avoid extra
+// dependencies and size.
+c10::IValue preprocess(
+    const Module& mod,
+    const c10::Dict<IValue, IValue>& method_compile_spec) {
+  return method_compile_spec;
+}
+#else
```

Comment:
Given that preprocess is used only on server side and we have removed preprocess from backend interface, then why should we have even the dummy preprocess if we are compiling delegate runtime?
Would it be possible or even make sense to refactor https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/backends/backend.h#L23?

Reply:
@kimishpatel and @SplitInfinity , good questions! I remember I discussed with @raziel before on the functions related to preprocess. There are two options. 1. to keep as is, but pass a dummy function or null on client side. 2. to move it out of backend.h. This PR is on option 1, but I could see the benefit of option 2. 

I could also see one pro of option2, to have backend registration in one line, including the (optional) preprocess. I remember there is also some backward compatibility needs for glow but I could remember wrong. 

For why I didn't use optional here is because the codes of preprocessing is in the same file of other APIs. Symbols in the server side preprocess can not be compiled. One option is to move preprocess into another file, including the registration part. 

@raziel could you confirm if there's any issue to move to option 2?

Reply:
The reason to have preprocess in backend.h is because it is used in the registration of a backend.

I think the discussion here is about registration on-host vs on-target, preprocess only exists on-host. 

I see these options:

1. Keep a single registration mechanism via the backend class: adding a constructor of the backend class that doesn't receive a preprocess function (I would prefer that to having an <optional> or null). We would just document such constructor for on-target registration.

2. Split the registration of preprocess and the backend delegation interface: remove the preprocess argument from the backend class constructor, and expose another API for detail::registerBackendPreprocessFunction(). That preprocess specific API/code should then live on a separate file such that we don't have do ifdef stuff (let me know if this doesn't make sense).

The advantage of (1) is that there's a single mechanism for backend registration, which means as a backend there's less chance for mistakes and you only care for linking backend.h. Caveat, which now I see is that we have to do the ugly macro thing because exactly you now bring a lot of baggage into backend.cpp

The advantage of (2) is that it's explicit what each registration is doing: one for interface, another for preprocessing, which makes particular sense for on-device cases where you really care about keeping those 2 worlds (and their dependencies) separate. That way it's explicit for a backend implementation that they're taking 2 separate dependencies depending on what they are building. Caveat is that there's now 2 things a backend will have to do to register their implementations, but in reality they'll have to do that anyway if they care to have a on-host and on-target builds (which I think many will). Plus, I think the extent of the mistakes that could happen is fairly small (forget to register preprocess or the interface, or mismatched backend names for the 2 registrations), but those should be pretty obvious to catch and fix.

So I suggest going with option #2 above.

Thoughts?

Reply:
@raziel I agree with option two. Although for ease of usage for backends that do not have distinct function for on-host and on-target, we can just have `detail::registerBackendWithPreprocessFunction` and `detail::registerBackend`. This way for cases like glow where compile and run is together the former API should suffice.

Reply:
@raziel and @kimishpatel , cool! I'll go with option 2, to split the preprocess registration from backend, and put both preprocess implementation and registration in a separate source file. 

I don't see any BC issues since all changes are on source code refactor, but will let you know if I find any. 

I also added a test on this PR, which will be updated soon.